crates/biome_js_analyze/src/lint/style/no_parameter_assign.rs

```diff
@@ -64,48 +116,142 @@ declare_lint_rule! {
+                                assignment.object().ok()
+                            }
+                            AnyJsAssignment::JsStaticMemberAssignment(assignment) => {
+                                assignment.object().ok()
+                            }
+                            _ => None,
+                        }
+                    })
+                    .map(ProblemType::PropertyAssignment)
+                    .collect();
+
+                signals.extend(expressions);
+            }
+
             if matches!(
                 declaration,
                 AnyJsBindingDeclaration::JsFormalParameter(_)
                     | AnyJsBindingDeclaration::JsRestParameter(_)
                     | AnyJsBindingDeclaration::JsArrowFunctionExpression(_)
                     | AnyJsBindingDeclaration::TsPropertyParameter(_)
             ) {
-                return binding.all_writes(ctx.model());
+                let param_reassignments: Vec<_> = binding
+                    .all_writes(model)
+                    .map(ProblemType::ParameterReassignment)
+                    .collect();
+
+                signals.extend(param_reassignments);
             }
         }
-        // Empty iterator that conforms to `AllBindingWriteReferencesIter` type.
-        std::iter::successors(None, |_| None)
+
+        signals
     }
 
-    fn diagnostic(ctx: &RuleContext<Self>, reference: &Self::State) -> Option<RuleDiagnostic> {
-        let param = ctx.query();
-        Some(
-            RuleDiagnostic::new(
-                rule_category!(),
-                reference.syntax().text_trimmed_range(),
-                markup! {
-                    "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
-                },
-            )
-            .detail(
-                param.syntax().text_trimmed_range(),
-                markup! {
-                    "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
-                },
-            )
-            .note(markup! {
-                "Use a local variable instead."
-            }),
-        )
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        match state {
+            ProblemType::ParameterReassignment(expression) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    expression.syntax().text_trimmed_range(),
+                    markup! {
+                        "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    },
+                )
+                    .detail(
+                        ctx.query().syntax().text_trimmed_range(),
+                        markup! {
+                        "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
+                    },
+                    )
+                    .note(markup! {
+                    "Use a local variable instead."
```

Comment:
We could take the opportunity of improving the diagnostic:

```suggestion
                    "Developers usually expect function parameters to be readonly. To align with this expectation, use a local variable instead."
```